gnovm/pkg/gnolang/alloc.go

```diff
@@ -96,15 +105,162 @@ func (alloc *Allocator) Fork() *Allocator {
 	}
 }
 
+func (alloc *Allocator) MemStats() string {
+	if alloc == nil {
+		return "nil allocator"
+	} else {
+		return fmt.Sprintf("Allocator{maxBytes:%d, bytes:%d}", alloc.maxBytes, alloc.bytes)
+	}
+}
+
+func (alloc *Allocator) GC() {
+	// using a throwaway allocator to recalculate memory usages
+	throwaway := NewAllocator(alloc.maxBytes, alloc.m)
+	throwaway.throwAway = true
+
+	// XXX, This is a rough implementation that
+	// estimates consumption based on the number
+	// of allocation cycles. A benchmark should
+	// be introduced for precise measurement, but
+	// that will be addressed in the future.
+	defer func() {
+		gasCPU := overflow.Mul64p(throwaway.allocTimes, GasFactorCPU)
+		if alloc.m.GasMeter != nil { //  no gas meter for test
+			alloc.m.GasMeter.ConsumeGas(gasCPU, "GC")
+		}
+	}()
+
+	// scan frames primarily to count occurrences of BoundMethod
+	for _, fr := range alloc.m.Frames {
+		if !fr.Receiver.IsUndefined() {
+			if pv, ok := fr.Receiver.V.(PointerValue); ok {
+				if _, ok := pv.Base.(*HeapItemValue); !ok {
+					// only alloc pointer for implicit
+					// address-taking, while using val
+					// invoke ptr method
+					throwaway.AllocatePointer()
+				} else {
+					throwaway.allocateValue(fr.Receiver.V)
+				}
+			}
+			if fr.Func != nil {
+				throwaway.allocateValue(fr.Func)
+			} else if fr.GoFunc != nil {
+				throwaway.allocateValue(fr.GoFunc)
+			}
+			// by default
+			throwaway.AllocateBoundMethod()
+		}
+	}
+
+	// scan blocks
+	for _, b := range alloc.m.Blocks {
+		throwaway.allocateValue(b)
+	}
+
+	// reset allocator
+	alloc.bytes = throwaway.bytes
+}
+
+func (alloc *Allocator) allocateTV(tv TypedValue) {
+	if tv.GetRefCount() < 2 {
```

Comment:
why < 2? e.g. what if == 2? 

I'm guessing it's a volatile counter. but it's a bit confusing because values already have persistent ref counts. maybe a rename is in order.

thinking..